src/goto-instrument/havoc_utils.cpp

```diff
@@ -62,4 +70,6 @@ void havoc_utilst::append_scalar_havoc_code_for_expr(
   side_effect_expr_nondett rhs(expr.type(), location);
   dest.add(goto_programt::make_assignment(
     code_assignt{expr, std::move(rhs), location}, location));
+  dest.instructions.back().source_location_nonconst().set_comment(
+    IS_LOOP_HAVOC);
```

Comment:
I appreciate that you want to put some kind of tag in the program but this feels fragile and ad-hoc.  Would it be possible to use a function call to `end_of_loop_havoc_block` or a `LOCATION` or even an `OTHER` instruction?  Without knowing the wider context (hard to get from such a big commit) I can't really say what is best.

Reply:
I removed `is_loop_havoc` function and related codes.
Instead, I changed the structure of transformed loop from
```
havoc(write_set)
```
to
```
DECL __is_in_loop_havoc_block
ASSIGN __is_in_loop_havoc_block = true
havoc(write_set)
ASSIGN __is_in_loop_havoc_block = false
```
Now, after applying a loop contract, `code_cotractst` will scan instrumented instructions to find all assignments in loop-havoc blocks and store them into a set `loop_havoc_set`.

@tautschnig : We cannot decide whether an assignment is loop havoc from its  instruction structure. So I change the structure of transformed loops to better identify loop havoc.